cli/src/command.rs

```diff
@@ -150,33 +150,172 @@ pub fn run() -> Result<()> {
+				})
+			} else if chain_spec.is_westend() {
+				runner.async_run(|mut config| {
+					let (client, _, _, task_manager) = service::new_chain_ops::<
+						service::westend_runtime::RuntimeApi,
+						service::WestendExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, config.database), task_manager))
+				})
+			} else {
+				runner.async_run(|mut config| {
+					let (client, _, _, task_manager) = service::new_chain_ops::<
+						service::polkadot_runtime::RuntimeApi,
+						service::PolkadotExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, config.database), task_manager))
+				})
+			}
+		},
+		Some(Subcommand::ExportState(cmd)) => {
+			let runner = cli.create_runner(cmd)?;
+			let chain_spec = &runner.config().chain_spec;
+
+			set_default_ss58_version(chain_spec);
+
+			if chain_spec.is_kusama() {
+				runner.async_run(|mut config| {
+					let (client, _, _, task_manager) = service::new_chain_ops::<
+						service::kusama_runtime::RuntimeApi,
+						service::KusamaExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, config.chain_spec), task_manager))
+				})
+			} else if chain_spec.is_westend() {
+				runner.async_run(|mut config| {
+					let (client, _, _, task_manager) = service::new_chain_ops::<
+						service::westend_runtime::RuntimeApi,
+						service::WestendExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, config.chain_spec), task_manager))
+				})
+			} else {
+				runner.async_run(|mut config| {
+					let (client, _, _, task_manager) = service::new_chain_ops::<
+						service::polkadot_runtime::RuntimeApi,
+						service::PolkadotExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, config.chain_spec), task_manager))
+				})
+			}
+		},
+		Some(Subcommand::ImportBlocks(cmd)) => {
+			let runner = cli.create_runner(cmd)?;
+			let chain_spec = &runner.config().chain_spec;
+
+			set_default_ss58_version(chain_spec);
+
+			if chain_spec.is_kusama() {
+				runner.async_run(|mut config| {
+					let (client, _, import_queue, task_manager) = service::new_chain_ops::<
+						service::kusama_runtime::RuntimeApi,
+						service::KusamaExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, import_queue), task_manager))
+				})
+			} else if chain_spec.is_westend() {
+				runner.async_run(|mut config| {
+					let (client, _, import_queue, task_manager) = service::new_chain_ops::<
+						service::westend_runtime::RuntimeApi,
+						service::WestendExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, import_queue), task_manager))
+				})
+			} else {
+				runner.async_run(|mut config| {
+					let (client, _, import_queue, task_manager) = service::new_chain_ops::<
+						service::polkadot_runtime::RuntimeApi,
+						service::PolkadotExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, import_queue), task_manager))
+				})
+			}
+		},
+		Some(Subcommand::PurgeChain(cmd)) => {
+			let runner = cli.create_runner(cmd)?;
+			runner.sync_run(|config| cmd.run(config.database))
```

Comment:
I'm not 100% sure but since we do have to repeat ourselves a lot, we'd better make it easier by using the same pattern everywhere. Is it possible to use `async_run` everywhere? Even if the code is blocking, it doesn't really matter that it in a future or not (little bit of overhead, that's ok).

Reply:
You need a `TaskManager` to pass into `async_run` (unless we make that an `Option<TaskManager>`?), and it seems like a waste to build one just for that.

Reply:
TaskManager?? :grin: That was not my idea when I made it. The run_node_until_exit was meant to use a task manager but sync_run and async_run where meant to run without task manager. Apparently it has been changed in https://github.com/paritytech/substrate/pull/6543

On the bright side: async_run seems not to be used at all anywhere except in this PR. Not even here in polkadot. So we can still change the course if we want.

It might be worth considering really only using a future in async_run and let the code inside handle any task manager they night have (and call clean_shutdown when the command needs to stop the node for example).